### DISCUSSION TITLE
Why does org.springFramework.Boot.loader.jar.JarFile Retained Heap continue to grow?

### DISCUSSION THREAD

**Discussion by CloudSen** [apache/skywalking#9436] [Q&A] [answered]
On Skywalking APM monitor, I notice one of my springboot server restart frequently, and jvm heap memory gradually grow.
![image](https://user-images.githubusercontent.com/25652685/183362086-729d5391-d746-45bc-9480-c221239dace8.png)

Then I exported multiple dump files, and use Eclipse MAT to analysis leaks, the results are shown in the following figure.
![image](https://user-images.githubusercontent.com/25652685/183363314-4a1136d5-cc90-4a1f-b666-3274c4c64f7e.png)
![image](https://user-images.githubusercontent.com/25652685/183363377-b4b0b009-9a83-472a-a0b4-d51939daa344.png)
![image](https://user-images.githubusercontent.com/25652685/183363409-3c4cf1cb-36db-4495-b1cd-47249d0c8fd9.png)

two dump compare:
![image](https://user-images.githubusercontent.com/25652685/183363643-73287b62-b4ee-4706-be84-7912f5318f1e.png)

GC Roots:
![image](https://user-images.githubusercontent.com/25652685/183363548-af943bf7-9e70-4488-9ab2-d66b0079b72f.png)

As you can see, Jarfile's retained heap has risen from 184m to 469m, and there are more and more objects. After one day of running, the size can reach 1.8GB!
Is this a bug about Skywalking or Spring Boot？
Im using Skywalking=>8.2.0 and SpringBoot=>2.3.3.release.

**Comment by wu-sheng** [comment, marked as answer]
1. 8.2.0 is a very old release.  It is more than one year old
2. Jar files should be unchanged unless you are using some tech to create new jar files, which we can't know.

**Reply by wu-sheng** [reply]
The typical way is you are using some kind of hot class generating mechanism in your codes. We don't know which, but it is possible.

**Reply by CloudSen** [reply]
Thx, I finally found the problem, That's because I used java.net.urlClassloader to dynamically load the jar package, then cached in Map.
By the way, I will try the new release version.
